### ISSUE TITLE
Cursor Style

### ISSUE THREAD

**Issue by BenQoo** [lvgl/lvgl#28]
Would it be possible to add additional cursor style?

for example:
* block style
![block](https://user-images.githubusercontent.com/29389982/30363218-f1c51a8e-9891-11e7-99e4-3bf0e710f42e.gif)
* block-outline style
![block-outline](https://user-images.githubusercontent.com/29389982/30363219-f1c59720-9891-11e7-9f7b-5c05a51edbcd.gif)
* underline style
![underline](https://user-images.githubusercontent.com/29389982/30363332-526aa3ae-9892-11e7-90d9-5c19e6def0c0.gif)

Thank you!

**Comment by kisvegabor** [comment]
Good idea, and there is no obstacle to implement it. I will do it soon!
Thank you!

**Comment by kisvegabor** [comment]
I started to implement this feature but some thing is not clear to me. How do these cursors should look like after the last character? (When you want to insert a char after last one) . There is no character to be framed.
Now I draw an empty frame or box like a terminal do?

Do you think is it required to specify cursor color different from text color?

**Comment by kisvegabor** [comment]
This is the current state (for some reasons it is blinking on the gif...):

![underline](https://user-images.githubusercontent.com/7599318/30379926-5adbf320-9898-11e7-820c-53ac02eebf58.gif)

![outline](https://user-images.githubusercontent.com/7599318/30379922-5947ab44-9898-11e7-8048-e04ce4827a7c.gif)

![line](https://user-images.githubusercontent.com/7599318/30379916-57a90a6c-9898-11e7-8df7-70621dd1e215.gif)

![block](https://user-images.githubusercontent.com/7599318/30379909-55b7fd76-9898-11e7-8c5c-358919a010eb.gif)

Any suggestions?

**Comment by BenQoo** [comment]
Hi,
> I started to implement this feature but some thing is not clear to me. How do these cursors should look like after the last character? (When you want to insert a char after last one) . There is no character to be framed.
> Now I draw an empty frame or box like a terminal do?

It looks great! 👍 
I check the windows console and Visual Studio Code editor, they also place cursor at the last empty character. I think it's no problem in current state.

![block](https://user-images.githubusercontent.com/29389982/30407484-0a6e2e7e-992b-11e7-81dd-765b1b2eb342.gif)
![block-outline](https://user-images.githubusercontent.com/29389982/30407486-0a7c8a82-992b-11e7-82bf-67fc6089ac1a.gif)
![underline](https://user-images.githubusercontent.com/29389982/30407485-0a761f12-992b-11e7-9f02-2f724d3dc0f2.gif)

> Do you think is it required to specify cursor color different from text color?

I have a concern about the block style cursor.
If cursor and character take the same color, it will be unable to distinguish character whenever cursor pops up. It might be solved by:
1. Change text color to reverse color or different color whenever cursor pops up.
2. Specify a cursor color which is different from text color and background color as your current design.

I'm not sure which one is the better implementation for littleV, but it will be more flexible for customization if the cursor has its own color.

**Comment by BenQoo** [comment]
Besides, I found 3 issues from the gif animation you posted.

1. The position of cursor is shifted left slightly.
Please refer to the following images:

   ![block 33](https://user-images.githubusercontent.com/29389982/30412870-4ca1329c-994d-11e7-9db6-d2dfcd6f02bb.png)  
   [block 33 enlarge](https://user-images.githubusercontent.com/29389982/30412873-52bfa3de-994d-11e7-8610-fb2ecb6488b7.png)

   ![block_line 7](https://user-images.githubusercontent.com/29389982/30412878-5c87e2d2-994d-11e7-896a-8976273a2378.png)
   [block_line 7 enlarge](https://user-images.githubusercontent.com/29389982/30412879-5c8abb06-994d-11e7-896b-b3c58f4361d8.png)

   ![block 34](https://user-images.githubusercontent.com/29389982/30412881-612c15d8-994d-11e7-9f86-90a8f2b6f8b4.png)
   [block 34 enlarge](https://user-images.githubusercontent.com/29389982/30412882-612e8318-994d-11e7-8fa8-f04a8f708439.png)

   ![block_line 8](https://user-images.githubusercontent.com/29389982/30412884-617eab40-994d-11e7-8a0c-bda8e909384b.png)
   [block_line 8 enlarge](https://user-images.githubusercontent.com/29389982/30412883-6158ac10-994d-11e7-8455-e05898ed086b.png)

1. The block cursor and block-line cursor don't popped up at whitespace character or it is just at the time within the invisible period of cursor(?).

1. I recommend to make cursor animation restart whenever edit text or move cursor. It will keep the cursor displayed until user stop inputting.
  ![add](https://user-images.githubusercontent.com/29389982/30414412-145e4484-9956-11e7-9c68-2709bd1a71d2.gif)
  ![move](https://user-images.githubusercontent.com/29389982/30414414-163ac5fc-9956-11e7-99bd-84da5b38c962.gif)

**Comment by kisvegabor** [comment]
Thank you for your feedback!

There was something bad with cursor position. Now I fixed it. 
I also reset the animation when a new cursor position is set.

Setting cursor color directly would be good ( e.g. lv_ta_set_cursor_color()) but I don't want to mix style properties with object attributes.  What do you think about set a cursor_style (lv_ta_set_cursor_style) to set:
* line_width: width in line, underline and outline mode
* mcolor: cursor colo
* ccolor: letter recolor in box mode

Adding a new styles is common among other object types too (like scroll bar style of page)

The current state is:
![block2](https://user-images.githubusercontent.com/7599318/30427779-62373a26-9951-11e7-9db3-2dc00cb2382b.gif)

**Comment by BenQoo** [comment]
Thanks for your reply!
I think it's no problem and the current state looks pretty nice! 👍

**Comment by kisvegabor** [comment]
I pushed the new feature into the beta branch. Can you try it?

`lv_ta_set_cursor_type(ta, LV_TA_CURSOR_BLOCK/LINE/OUTLINE/UNDERLINE)` will modify the cursor type.

`lv_ta_set_cursor_style(ta, &style)` will set a new style.
* line_width: UNDERLINE and LINE width
* mcolor, gcolor: BLOCK, UNDERLINE, LINE color
* bcolor: OUTLINE color
* bwidth: OUTLINE width

By default the cursor style is create according to the label's style.

**Comment by BenQoo** [comment]
Sorry for my late reply.
I will have time to test it next week.
Thanks for your update!

**Comment by kisvegabor** [comment]
Okay, take your time! :)

**Comment by BenQoo** [comment]
I found some issues and list as follows:

#### Line style
1. Cursor width and color will be overwritten by `bwidth` and `bcolor`
  [screenshot](https://user-images.githubusercontent.com/29389982/30791713-e968ee20-a1e6-11e7-9d22-c70634360ed8.png)
   ```C
    cursor_style.line_width = 3 * LV_DOWNSCALE; // UNDERLINE and LINE width
    cursor_style.mcolor = COLOR_RED;            // BLOCK, UNDERLINE, LINE color
    cursor_style.gcolor = COLOR_YELLOW;         // BLOCK, UNDERLINE, LINE color
    cursor_style.bcolor = COLOR_BLUE;           // OUTLINE color
    cursor_style.bwidth = 6 * LV_DOWNSCALE;     // OUTLINE width
   ```

#### UnderLine style
1. Same as _Line style_ issue 1.
1. Cursor draw abnormally at `\n` character. (only draw one pixel column)
   [screenshot](https://user-images.githubusercontent.com/29389982/30791760-5021f6fc-a1e7-11e7-8d3f-fea7398ece84.png)
   ```C
    cursor_style.line_width = 6 * LV_DOWNSCALE; // UNDERLINE and LINE width
    cursor_style.mcolor = COLOR_RED;            // BLOCK, UNDERLINE, LINE color
    cursor_style.gcolor = COLOR_YELLOW;         // BLOCK, UNDERLINE, LINE color
    lv_ta_set_text(ta_1, "123 56\nAB EFG");
    lv_ta_set_cursor_pos(ta_1, 6);
   ```
1. Cursor will cover character if set a larger `line_width`.
   [screenshot](https://user-images.githubusercontent.com/29389982/30800924-2d460376-a214-11e7-8067-cf4770d65866.png)
    I recommend to take a `vpad` in cursor style to adjust the width of gap between underline cursor and character, or increase the thickness of cursor toward the bottom side?

#### Outline style
1. Cursor draw abnormally at `\n` character.
    [screenshot](https://user-images.githubusercontent.com/29389982/30793281-69e001b8-a1f3-11e7-8171-f31d738b85cc.png)
   ```C
    cursor_style.bcolor = COLOR_BLUE;           // OUTLINE color
    cursor_style.bwidth = 6 * LV_DOWNSCALE;     // OUTLINE width
    lv_ta_set_text(ta_1, "123 56\nAB EFG");
    lv_ta_set_cursor_pos(ta_1, 6);
   ```

#### Block style
1. Cursor draw abnormally at `\n` character. (only draw one pixel column)

**Comment by kisvegabor** [comment]
Hi,
thank you for this deep testing!

You are right with incorrect drawing on `\n`. I fixed it.

LINE, UNDERLINE and BLOCK style are drawn like a rectangle. So if you set a very thick border then it will occupy the whole space. Try `bwidth = 1 * LV_DOWNSCALE` on this styles. Sorry, it wasn't clear when I described the meaning of the style fields. So:
* line_width: UNDERLINE and LINE width
* mcolor, gcolor: BLOCK, UNDERLINE, LINE color
* bcolor: OUTLINE color *and BLOCK, UNDERLINE, LINE border color* 
* bwidth: OUTLINE width *and BLOCK, UNDERLINE, LINE border width*

I defined the style this way:
```c
static lv_style_t cursor_style;
lv_style_get(LV_STYLE_PLAIN, &cursor_style);
cursor_style.line_width = 3 * LV_DOWNSCALE; // UNDERLINE and LINE width
cursor_style.mcolor = COLOR_RED;            // BLOCK, UNDERLINE, LINE color
cursor_style.gcolor = COLOR_YELLOW;         // BLOCK, UNDERLINE, LINE color
cursor_style.bcolor = COLOR_BLUE;           // OUTLINE color and BLOCK, UNDERLINE, LINE border color
cursor_style.bwidth = 1 * LV_DOWNSCALE;     // OUTLINE color and BLOCK, UNDERLINE, LINE border color
cursor_style.ccolor = COLOR_LIME;           // BLOCK letter re-color
cursor_style.opa = OPA_50;                  //Cursor opacity
```

It looks like this from me right now:
![screenshot from 2017-09-25 12-20-06](https://user-images.githubusercontent.com/7599318/30803967-f1a0e476-a1eb-11e7-9b7e-421b33e264d2.png)

**Comment by BenQoo** [comment]
Hi,
Thanks for your explanation, I get it.

I have a additional question about UNDERLINE cursor.
Currently, UNDERLINE cursor increase the thickness toward the top(character) side.The room between UNDERLINE cursor and character will become narrow if chose a smaller font size.

![example](https://user-images.githubusercontent.com/29389982/30838863-b1edbca4-a2a0-11e7-863a-4e4826233731.png)

Such as the above screenshot, it only have 1 pixel height to increase UNDERLINE width.
Is it possible to adjust padding between them or make the thickness increased toward the bottom side?

Thank you!

**Comment by kisvegabor** [comment]
It is a good point! A can adjust it equally up and down. 
Maybe the same thing should be done with LINE type but left and right directions.

**Comment by kisvegabor** [comment]
Now I align LINE and UNDERLINE equally top-bottom/left-right.

**Comment by BenQoo** [comment]
I found a concern about d6ccc48963343119708b032d046d54db9bf4ff68.
To increase the width toward the both side, It will have the same problem as the original design.

I have an idea which is to take cursor style's `vpad` and `hpad` to adjust the cursor position, 
and we don't have to consider the directions for width increasing.
What do you think?

```C
if (ta_ext->cursor_type == LV_TA_CURSOR_LINE) {
    cur_area.x1 = letter_pos.x + ta_ext->label->cords.x1 + cur_style.hpad;
    cur_area.y1 = letter_pos.y + ta_ext->label->cords.y1 + cur_style.vpad;
    cur_area.x2 = letter_pos.x + ta_ext->label->cords.x1 + cur_style.hpad + cur_style.line_width;
    cur_area.y2 = letter_pos.y + ta_ext->label->cords.y1 + cur_style.vpad + letter_h;
    lv_draw_rect(&cur_area, mask, &cur_style);
} else if (ta_ext->cursor_type == ...) {
    ...
} else if (ta_ext->cursor_type == LV_TA_CURSOR_UNDERLINE) {
    cur_area.x1 = letter_pos.x + ta_ext->label->cords.x1 + cur_style.hpad;
    cur_area.y1 = letter_pos.y + ta_ext->label->cords.y1 + cur_style.vpad + letter_h - cur_style.line_width;
    cur_area.x2 = letter_pos.x + ta_ext->label->cords.x1 + cur_style.hpad + letter_w;
    cur_area.y2 = letter_pos.y + ta_ext->label->cords.y1 + cur_style.vpad + letter_h;
    lv_draw_rect(&cur_area, mask, &cur_style);
}
```

**Comment by kisvegabor** [comment]
Added as you have suggested.
In addition OUTLINE and BLOCK size will increased with +-hpad/vpad.

**Comment by BenQoo** [comment]
Hi,
There is an issue about outline cursor, `vpad` didn't change the size, it change the position.

Besides, the line of border is blurred at some edge, is it the expected rendering?
![screenshot](https://user-images.githubusercontent.com/29389982/31596570-e91eb4b4-b275-11e7-8c53-bcc26cfdb0ac.png)

**Comment by kisvegabor** [comment]
Hi,
a sign was bad. I fixed it in `beta`.

The blur is the expected behavior because of anti aliasing. When a 1 px wide line should be drawn between two pixels then two lighter line will appear.

**Comment by BenQoo** [comment]
Thanks for your reply, I get it.

`vpad` on outline cursor is working now. :)

**Comment by kisvegabor** [comment]
It seems I'm not so good at testing things very well. :) To work with less bug I decided to create test codes for all object types. Unfortunately it can happen that a new feature breaks an old or "something" happens during merge. All test cases will contain images too about the expected result.

It will be launched in the next released. The current state is here: https://github.com/littlevgl/examples/tree/dev-5.0/lv_tests

**Comment by BenQoo** [comment]
👍 

However, I think it's common that it could meet some issues or bugs while developing a new feature. so don't take that too much to heart. :)

**Comment by kisvegabor** [comment]
I released the beta version of v5.0 (lvgl/beta, misc/beta, proj_pc/beta).
As it is major version it is not backward compatible. (mainly because of the styles)
This small code shows the new style usage system:
```c
    static lv_style_t ta_cur_style;
    lv_style_copy(&ta_cur_style, &lv_style_pretty);
    ta_cur_style.body.main_color = COLOR_RED;

    lv_obj_t *ta = lv_ta_create(lv_scr_act(), NULL);
    lv_ta_set_cursor_type(ta, LV_CURSOR_BLOCK);
    lv_ta_set_style(ta, LV_TA_STYLE_CURSOR, &ta_cur_style);
```
I hope you can make some test to approve this feature in v5.0.
(Please update you `misc_conf.h` and` lv_conf.h` from `misc_conf_templ.h` and` lv_conf_templ.h`)

**Comment by kisvegabor** [comment]
Hi,

thank you for raising the idea and for testing.
This feature is finally released in  v5.0.0
